src/tree/updater_quantile_hist.cc

```diff
@@ -680,30 +683,63 @@ bool QuantileHistMaker::Builder<GradientSumT>::UpdatePredictionCache(
     }
   });
 
+  if (param_.subsample < 1.0f) {
```

Comment:
Is it potentially easier to complete this step at the end of the update instead of in `UpdatePredictionCache`? That way we don't need p_last_fmat_mutable_ as we are guaranteed to have a valid DMatrix.

Reply:
I don't think it will be easier. We already have other things in `UpdatePredictionCache()` needed for predicting, such as `out_preds`, the number of trees, and the number of the current tree. Moving them to `Update()` will also cost us a few extra fields in Updater. But we can probably remove `const` from the existing `p_last_fmat_` field.

Reply:
Ok that seems fair. I don't like this `p_last_fmat_` business much - it's a raw pointer that we are making assumptions about. For example if someone made changes to other parts of the program it could be easy to invalidate its use in the updater in unpredictable situations. If you can think of a better way to handle this in general that could be a nice change for the future.

Reply:
Thank you for sharing your vision. We are going to do some refactoring for `hist` in the nearest future. At first glance, we can replace raw pointers with smart ones. Next, perhaps, we will think about some hashing for DMatrix.

Reply:
You should definitely chat with @trivialfis about `hist` refactoring and work on an RFC.